File: Cpp/71.简化路径.cpp
/*
 * @lc app=leetcode.cn id=71 lang=cpp
 *
 * [71] 简化路径
 */

#include <iostream>
#include <string>
#include <queue>

using namespace std;

// @lc code=start
class Solution {
public:
    string simplifyPath(string path) {
        deque<string> paths;
        int n = (int)path.size();
        for (int i = 1; i < n;) {
            if (path[i] == '/' && ++i >= 0) continue;
            int j = i + 1;
            while (j < n && path[j] != '/') ++j;
            string item = path.substr(i, j - i);
            if (item == "..") {
                if (!paths.empty()) paths.pop_back();
            } else if (item != ".") {
                paths.push_back(item);
            }
            i = j;
        }
        string ans = "";
        while (!paths.empty()) {
            ans += "/" + paths.front();
            paths.pop_front();
        }
        return ans.size() == 0 ? "/" : ans;
    }
};
// @lc code=end

int main(int argc, const char * argv[]) {
    // insert code here...
    
    string path;
    string res;
    
    /**
     输入：path = "/home/"
     输出："/home"
     解释：注意，最后一个目录名后面没有斜杠。
     */
    path = "/home/";
    res = "/home";
    assert(res == Solution().simplifyPath(path));
    
    /**
     输入：path = "/../"
     输出："/"
     解释：从根目录向上一级是不可行的，因为根目录是你可以到达的最高级。
     */
    path = "/../";
    res = "/";
    assert(res == Solution().simplifyPath(path));
    
    /**
     输入：path = "/home//foo/"
     输出："/home/foo"
     解释：在规范路径中，多个连续斜杠需要用一个斜杠替换。
     */
    path = "/home//foo/";
    res = "/home/foo";
    assert(res == Solution().simplifyPath(path));
    
    /**
     输入：path = "/a/./b/../../c/"
     输出："/c"
     */
    path = "/a/./b/../../c/";
    res = "/c";
    assert(res == Solution().simplifyPath(path));

    cout << "OK~" << endl;

    return 0;
}

File: Swift/71.简化路径.swift
/*
 * @lc app=leetcode.cn id=71 lang=swift
 *
 * [71] 简化路径
 */

// @lc code=start
class Solution {
    func simplifyPath(_ path: String) -> String {
        let paths = path.map { $0 }
        var cpns: [String] = []
        var i = 1   // "/a//b////c/d//././/.."
        while i < paths.count {
            if paths[i] == "/" {
                i += 1
                continue
            }
            var j = i + 1
            while j < paths.count && paths[j] != "/" {
                j += 1
            }
            let item = String(paths[i..<j])
            if item == ".." {
                if !cpns.isEmpty {
                    cpns.removeLast()
                }
            } else if item != "." {
                cpns.append(item)
            }
            i = j
        }
        var ans = ""
        while !cpns.isEmpty {
            ans += "/" + cpns.removeFirst()
        }
        return ans.isEmpty ? "/" : ans
    }
}
// @lc code=end

func main() {
    var path: String
    var res: String
    
    /**
     输入：path = "/home/"
     输出："/home"
     解释：注意，最后一个目录名后面没有斜杠。
     */
    path = "/home/"
    res = "/home"
    assert(res == Solution().simplifyPath(path))

    /**
     输入：path = "/../"
     输出："/"
     解释：从根目录向上一级是不可行的，因为根目录是你可以到达的最高级。
     */
    path = "/../"
    res = "/"
    assert(res == Solution().simplifyPath(path))

    /**
     输入：path = "/home//foo/"
     输出："/home/foo"
     解释：在规范路径中，多个连续斜杠需要用一个斜杠替换。
     */
    path = "/home//foo/"
    res = "/home/foo"
    assert(res == Solution().simplifyPath(path))

    /**
     输入：path = "/a/./b/../../c/"
     输出："/c"
     */
    path = "/a/./b/../../c/"
    res = "/c"
    assert(res == Solution().simplifyPath(path))
    
    /**
     输入：path = "/a//b////c/d//././/.."
     输出："/a/b/c"
     */
    path = "/a//b////c/d//././/.."
    res = "/a/b/c"
    assert(res == Solution().simplifyPath(path))
}
